Adds model$format()

rok-cesnovar opened the pull request:
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and agree to license (see below)

#### Summary

Fixes #624 

Example to run: 

```r
library(cmdstanr)

model_file <- write_stan_file("
data {
  int<lower=0> N;
  int<lower=0> n_redcards[N];
  int<lower=0> n_games[N];
  vector[N] rating;
}
parameters {
  vector[2] beta;
}


model {
  beta[1] ~
  normal(0, 10);
  beta[2] ~         normal(0, 1);


  target += binomial_logit_log(n_redcards, n_games, beta[1] + beta[2] * rating);
}
")

mod <- cmdstan_model(model_file, compile = FALSE)

# will print short lines
mod$format(max_line_length = 50, canonicalize = list("deprecations")) 
# will print longer lines
mod$format(max_line_length = 150, canonicalize = list("deprecations"))
# will write to the file
mod$format(max_line_length = 150, canonicalize = list("deprecations"), overwrite_file = TRUE, backup = TRUE)
file.exists(paste0(mod$stan_file(), ".bak")) # should be TRUE

file.remove(paste0(mod$stan_file(), ".bak"))
mod$format(max_line_length = 150, canonicalize = list("deprecations"), overwrite_file = TRUE, backup = TRUE)
file.exists(paste0(mod$stan_file(), ".bak")) # should be FALSE
```

The currently allowed `canonicalize` options are `deprecations`, `parentheses`, `braces` and `includes`.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting 
(this will be you or your assignee, such as a university or company): 
Rok Češnovar


By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the following licenses:

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)    
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

rok-cesnovar commented:
@jgabry when you have a minute please take a look. If you agree with this new method, I will just add a few tests and then should be good to go. This is mostly just trying to follow the lead of https://github.com/stan-dev/cmdstanpy/pull/539

jgabry commented on R/model.R:
```diff
@@ -767,6 +767,127 @@ check_syntax <- function(pedantic = FALSE,
+                   max_line_length = NULL) {
+  if (length(self$stan_file()) == 0) {
+    stop(
+      "'$check_syntax()' cannot be used because the 'CmdStanModel'",
```
```suggestion
      "'$format()' cannot be used because the 'CmdStanModel'",
```

jgabry commented on R/model.R:
```diff
@@ -767,6 +767,127 @@ check_syntax <- function(pedantic = FALSE,
+#'   to the input model file. The default is `FALSE`.
+#' @param canonicalize (list or logical) Defines whether or not the compiler
+#'   should 'canonicalize' the Stan model, removing things like deprecated syntax.
+#'   Default is `FALSE``. If `TRUE`, all canonicalizations are run. You can also
```
```suggestion
#'   Default is `FALSE`. If `TRUE`, all canonicalizations are run. You can also
```

jgabry commented on R/model.R:
```diff
@@ -767,6 +767,127 @@ check_syntax <- function(pedantic = FALSE,
+#' @family CmdStanModel methods
+#'
+#' @description The `$format()` method of a [`CmdStanModel`] object
+#'   run the stanc's auto-formatter on the model code. Either saves the formatted
```
```suggestion
#'   runs an auto-formatter on the model code. Either saves the formatted
```

jgabry reviewed (commented):
Looks great, just a few minor edits suggested

rok-cesnovar commented:
Added tests, should be good to go.

WardBrian commented:
In CmdStanPy I do some jumping through hopes to support versions pre-2.29 (basically preventing bad arguments from being passed). Not sure if you want to do this as well

WardBrian commented:
In Python we do the following:

Throw an error if version < 2.29 and max line length is set
Throw an error if version < 2.29 and canonicalize is set to anything other than 'True'
Make sure that if version is <2.29, at most one of '--print-canonical' and '--auto-format' is set (previously if you set both, stanc printed the program _twice_)

spinkney commented:
This is exciting!

rok-cesnovar commented:
> Throw an error if version < 2.29 and max line length is set
Throw an error if version < 2.29 and canonicalize is set to anything other than 'True'
Make sure that if version is <2.29, at most one of '--print-canonical' and '--auto-format' is set (previously if you set both, stanc printed the program twice)

Thanks! Will add this right now.

rok-cesnovar commented:
A couple modifications: 
- added the version checks Brian suggested,
- added timestamp to the backup file
- if backed up, the path to the backup file is printed in a message
- added a quiet arg to not print the message

Also added some more tests.

jgabry reviewed (approved):
Thanks! Ready to merge if all the checks pass.
